packages/widget.client-react/src/context.ts

```diff
@@ -15,22 +15,76 @@
  */
 
 import type { ObjectSet } from "@osdk/client";
-import type { ObjectType } from "@osdk/widget.api";
+import type {
+  EventId,
+  EventParameterValueMap,
+  ObjectType,
+  WidgetMessage,
+} from "@osdk/widget.api";
 import {
   type AsyncParameterValueMap,
   type AsyncValue,
   FoundryHostEventTarget,
-  type FoundryWidgetClient,
   type ParameterConfig,
   type ParameterValueMap,
   type WidgetConfig,
 } from "@osdk/widget.client";
 import React, { useContext } from "react";
 
+export type AugmentedEventParameterValueMap<
+  C extends WidgetConfig<C["parameters"]>,
+  K extends EventId<C>,
+> = {
+  [P in keyof EventParameterValueMap<C, K>]: P extends keyof C["parameters"]
+    ? C["parameters"][P] extends { type: "objectSet"; objectType: infer T }
+      ? T extends ObjectType ? ObjectSet<T>
+      : EventParameterValueMap<C, K>[P]
+    : EventParameterValueMap<C, K>[P]
+    : EventParameterValueMap<C, K>[P];
+};
+
+type AugmentedEmitEventIdMap<C extends WidgetConfig<C["parameters"]>> = {
+  [K in EventId<C>]: {
+    eventId: K;
+    parameterUpdates: AugmentedEventParameterValueMap<C, K>;
+  };
+};
+
+interface AugmentedWidgetEmitEvent<C extends WidgetConfig<C["parameters"]>>
+  extends Omit<WidgetMessage.EmitEvent<C>, "payload">
+{
+  payload: AugmentedEmitEventIdMap<C>[EventId<C>];
+}
+
+/**
+ * Emits a widget event with parameter updates to the host Foundry UI.
+ *
+ * For object set parameters, pass `ObjectSet<T>` directly.
+ * If the same event is emitted multiple times in quick succession, only the last call will be sent.
+ *
+ * Errors (e.g., network failures when materializing ObjectSet parameters)
```

Comment:
let's avoid confusion with other meaning of materialize object set
```suggestion
 * Errors (e.g., network failures when serializing ObjectSet parameters)
```